src/coreclr/nativeaot/System.Private.CoreLib/src/System/Attribute.NativeAot.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime;
+using System.Runtime.CompilerServices;
+
+using Internal.Runtime;
+
+namespace System
+{
+    public abstract partial class Attribute
+    {
+        // An override of this method will be injected by the compiler into all attributes that have fields.
+        // This API is a bit awkward because we want to avoid burning more than one vtable slot on this.
+        // When index is negative, this method is expected to return the number of fields of this
+        // valuetype. Otherwise, it returns the offset and type handle of the index-th field on this type.
+        internal virtual unsafe int __GetFieldHelper(int index, out MethodTable* mt)
+        {
+            Debug.Assert(index < 0);
+            mt = default;
+            return 0;
+        }
+
+        public override unsafe bool Equals([NotNullWhen(true)] object? obj)
+        {
+            if (obj == null)
+                return false;
+
+            if (this.GetType() != obj.GetType())
+                return false;
+
+            int numFields = __GetFieldHelper(-1, out _);
+
+            // Subtract nint size to get to the m_pEEType field. Can't get from "ref MethodTable*" to "ref byte" currently.
+            ref byte thisRawData = ref Unsafe.Subtract(ref this.GetRawData(), nint.Size);
```

Comment:
byref pointing at offset 0 is not quite right (it happens to work in the current runtime). It would be better to adjust the offset.